### PR TITLE
Group Normalization

### PR DESCRIPTION
**Pull request by haojin2**
## Description ##
[Group Normalization](https://arxiv.org/abs/1803.08494)

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [x] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [x] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [x] Forward
- [x] Backward
- [x] Unit tests
- [x] Gluon block
- [x] Unit test for Gluon

## Comments ##
Related:
#14196 #11283 
Flakiness Checks:
```
MXNET_TEST_COUNT=10000 nosetests tests/python/unittest/test_operator.py:test_groupnorm
[INFO] Setting module np/mx/python random seeds, use MXNET_MODULE_SEED=1680041162 to reproduce.
.
----------------------------------------------------------------------
Ran 1 test in 136.546s

OK
```
```
MXNET_TEST_COUNT=10000 nosetests tests/python/gpu/test_operator_gpu.py:test_groupnorm
[INFO] Setting module np/mx/python random seeds, use MXNET_MODULE_SEED=2085401116 to reproduce.
.
----------------------------------------------------------------------
Ran 1 test in 172.874s

OK
```

### CODE REVIEW

**Comment by haojin2** [comment]
@junrushao1994 For review if you have time.

**Comment by szha** [comment]
cc @sxjscience

**Comment by hetong007** [comment]
@zhanghang1989 FYI

File: src/operator/nn/group_norm-inl.h [hunk 1]
```diff
@@ -0,0 +1,317 @@
...
+    data_size = sizeof(DType) * data.Size();
+    red_out_size = sizeof(DType) * mean.Size();
+    // There are two types of reduction workloads: reduce over axis and reduce exclude axis
+    // We take the maximum of the workspace sizes required by these workloads.
+    // Also, we explicitly set the req_type=kAddto in case we want to use it.
+    BROADCAST_NDIM_SWITCH(red_dst_shape.ndim(), NDim, {
+      reduce_workspace_size =
+        std::max(reduce_workspace_size,
+                 broadcast::ReduceWorkspaceSize<NDim, DType>(s, red_src_shape,
+                                                             kAddTo, red_dst_shape));
```
Comment (sxjscience) [inline src/operator/nn/group_norm-inl.h]: Should switch the order between `red_dst_shape` and `red_src_shape`. See https://github.com/apache/incubator-mxnet/pull/14870/files ... I previous made that error. 😩😩

File: src/operator/nn/group_norm-inl.h [hunk 2]
```diff
@@ -0,0 +1,317 @@
...
+    });
+    BROADCAST_NDIM_SWITCH(red_exclude_dst_shape.ndim(), NDim, {
+      reduce_workspace_size =
+        std::max(reduce_workspace_size,
+                 broadcast::ReduceWorkspaceSize<NDim, DType>(s, red_exclude_src_shape, kAddTo,
+                                                             red_exclude_dst_shape));
```
Comment (sxjscience) [inline src/operator/nn/group_norm-inl.h]: Same here.

File: src/operator/nn/group_norm-inl.h (hunk 1 above)
Comment (haojin2) [reply src/operator/nn/group_norm-inl.h]: Done.

File: src/operator/nn/group_norm-inl.h (hunk 2 above)
Comment (haojin2) [reply src/operator/nn/group_norm-inl.h]: Done.

File: python/mxnet/gluon/nn/basic_layers.py [hunk 3]
```diff
@@ -616,6 +617,94 @@ def __repr__(self):
...
+    scale: bool, default True
+        If True, multiply by `gamma`. If False, `gamma` is not used.
+    beta_initializer: str or `Initializer`, default 'zeros'
+        Initializer for the beta weight.
+    gamma_initializer: str or `Initializer`, default 'ones'
+        Initializer for the gamma weight.
+
+
+    Inputs:
+        - **data**: input tensor with arbitrary shape.
```
Comment (sxjscience) [inline python/mxnet/gluon/nn/basic_layers.py]: It's not arbitrary shape. The input tensor needs to have shape (N, C, ...)

Comment (sxjscience) [inline python/mxnet/gluon/nn/basic_layers.py]: It's not arbitrary shape. The input tensor needs to have shape (N, C, ...)

File: src/operator/nn/group_norm.cc [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+Normalizes the channels of the input tensor by mean and variance, and applies a scale ``gamma`` as
+well as offset ``beta``.
+
+Assume the input has more than one dimension and we normalize along axis 1.
+We first compute the mean and variance along this axis and then
+compute the normalized output, which has the same shape as input, as following:
+
+.. math::
+
+  out = \frac{data - mean(data, axis)}{\sqrt{var(data, axis) + \epsilon}} * gamma + beta
```
Comment (sxjscience) [inline src/operator/nn/group_norm.cc]: Change the docstring here to reflect the computation logic of group norm, i.e.,

```
data = data.reshape((N, num_groups, C // num_groups, ...))
out = \frac{data - mean(data, 2)}{\sqrt{var(data, 2) + \epsilon}} * gamma + beta
```

Comment (haojin2) [reply src/operator/nn/group_norm.cc]: Done.

File: python/mxnet/gluon/nn/basic_layers.py (hunk 3 above)
Comment (haojin2) [reply python/mxnet/gluon/nn/basic_layers.py]: Done.

Comment (haojin2) [reply python/mxnet/gluon/nn/basic_layers.py]: Duplicate review.

File: src/operator/nn/group_norm.cc [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+  return true;
+}
+
+NNVM_REGISTER_OP(GroupNorm)
+.describe(R"code(Group normalization.
+
+Normalizes the channels of the input tensor by mean and variance, and applies a scale ``gamma`` as
+well as offset ``beta``.
+
+Assume the input has more than one dimension and we normalize along axis 1.
```
Comment (zhreshold) [inline src/operator/nn/group_norm.cc]: This assumption can confuse users. IMO pytorch doc is easier to understand: https://pytorch.org/docs/stable/nn.html#groupnorm

Comment (haojin2) [reply src/operator/nn/group_norm.cc]: Forgot to change the doc when I copied from LayerNorm...Working on that now.

Comment (haojin2) [reply src/operator/nn/group_norm.cc]: Done.

**Comment by zhreshold** [comment]
@haojin2 can you check why unitest is failing?

**Comment by haojin2** [comment]
@zhreshold Looking into it...

**Comment by haojin2** [comment]
@zhreshold The failure is on fp16 data type, previously it could endure 10000 trials on my end and now it could not, I'm looking into the issue.

**Comment by zhreshold** [comment]
@haojin2 is CI still failing?

**Comment by Jerryzcn** [comment]
Is there an issue related to CI failing?

**Comment by karan6181** [comment]
@sxjscience @zhreshold Could you please review this PR again?

@haojin2 Could you please re-trigger the CI and look into failure ? Thanks

**Comment by haojin2** [comment]
@zhreshold @Jerryzcn @sxjscience @eric-haibin-lin CI passed, ready for merge

**Comment by sxjscience** [comment]
I'll merge first. Need to make a PR to revise the test_util + add test for grad_req=kAddTo. Let's make sure all OPs are tested for grad_req=kAddTo (Need to combine the efforts with DeepNumpy).

**Pull request merged**